infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -141,280 +158,315 @@ private boolean isBootstrapped(CorfuMsg msg, ChannelHandlerContext ctx, IServerR
+                        requestHeader.getRequestId(), layout);
+
+                responseHeader = getHeaderMsg(requestHeader, false, false);
+                response = getResponseMsg(responseHeader,
+                        getBootstrapLayoutResponseMsg(BootstrapLayoutResponseMsg.Type.NACK));
+            } else {
                 setCurrentLayout(layout);
                 serverContext.setServerEpoch(layout.getEpoch(), r);
-                //send a response that the bootstrap was successful.
-                r.sendResponse(ctx, msg, new CorfuMsg(CorfuMsgType.ACK));
+
+                responseHeader = getHeaderMsg(requestHeader, false, true);
+                response = getResponseMsg(responseHeader,
+                        getBootstrapLayoutResponseMsg(BootstrapLayoutResponseMsg.Type.ACK));
             }
         } else {
-            // We are already bootstrapped, bootstrap again is not allowed.
-            log.warn("handleMessageLayoutBootstrap: Got a request to bootstrap a server which is "
-                    + "already bootstrapped, rejecting!");
-            r.sendResponse(ctx, msg, new CorfuMsg(CorfuMsgType.LAYOUT_ALREADY_BOOTSTRAP));
+            log.warn("handleBootstrapLayoutRequest[{}]: Got a request to bootstrap a server which is "
+                    + "already bootstrapped, rejecting!", requestHeader.getRequestId());
+
+            responseHeader = getHeaderMsg(requestHeader, false, true);
+            response = getResponseMsg(responseHeader, getBootstrappedErrorMsg());
         }
+
+        r.sendResponse(response, ctx);
     }
 
     /**
      * Accepts a prepare message if the rank is higher than any accepted so far.
      *
-     * @param msg corfu message containing LAYOUT_PREPARE
+     * @param req corfu message containing PREPARE_LAYOUT_REQUEST
      * @param ctx netty ChannelHandlerContext
      * @param r   server router
      */
-    // TODO this can work under a separate lock for this step as it does not change the global
-    // components
-    @ServerHandler(type = CorfuMsgType.LAYOUT_PREPARE)
-    public synchronized void handleMessageLayoutPrepare(
-            @NonNull CorfuPayloadMsg<LayoutPrepareRequest> msg,
-            ChannelHandlerContext ctx,
-            @NonNull IServerRouter r) {
-
-        // Check if the prepare is for the correct epoch
-        if (!isBootstrapped(msg, ctx, r)) {
-            return;
-        }
+    // TODO this can work under a separate lock for this step as it does not change the global components
+    @RequestHandler(type = RequestPayloadMsg.PayloadCase.PREPARE_LAYOUT_REQUEST)
+    public synchronized void handlePrepareLayoutRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        final HeaderMsg requestHeader = req.getHeader();
+        final PrepareLayoutRequestMsg payload = req.getPayload().getPrepareLayoutRequest();
+
+        if(!isBootstrapped(requestHeader, ctx, r)) return;
```

Comment:
I no you didn't write this, but I think its cleaner to make the `isBootstrapped` such that it doesn't have a side effect. Maybe somethign like this:

```java
if(!isBootstrapped(requestHeader, ctx, r)) {
   r.sendNoBootstrapError(reqHeader, ctx);
}
```